### ISSUE TITLE
Route Parameter ModelBind behaviour change in 2.1

### ISSUE THREAD

**Issue by stevejgordon** [aspnet/Mvc#7413]
I've attempted an upgrade of an ASP.NET Core 2.0 project to 2.1 using the preview1 packages now available on nuget.
 
We have a custom IModelBinderProvider in the app and I noticed an exception being thrown since the upgrade. I noticed this on a GET action which accepts a route parameter.

In 2.0 the route parameter is ignored by our IModelBinderProvider as the Metadata.PropertyName is null.
 
In 2.1 a PropertyName is now set for the incoming route paramter but the ContainerType is null. 
 
At the moment this blows up in our code. Is this in intended behaviour change?
 
I've thrown together a rough repro at https://github.com/stevejgordon/ModelBindRepro
 
Basically, start the 2.0 app and hit /Home/1 - This works fine. If you debug line 15 of the AdjustToTimezoneModelBinderProvider you'll see that PropertyName is null.
 
Start the 2.1 app and hit /Home/1 - An exception occurs at line 24 of AdjustToTimezoneModelBinderProvider because ContainerType is null. PropertyName is set to "statusCode". Also note that if you hover over Metadata there is a debugger null ref exception too since ModelMetadata.DebuggerToString() also tries to access the ContainerType.
 
I've not had time to go any deeper at this point.

//cc @rynowak, @mkArtakMSFT and @DamianEdwards

**Comment by mkArtakMSFT** [comment]
Thanks @stevejgordon. @dougbu, can you please look into this?

**Comment by dougbu** [comment]
2.1 Preview1 adds support for `ValidationAttribute`s on action parameters and controller properties. That new feature includes a new `ModelMetadataKind.Parameter` and, for that kind, `ModelMetadata.PropertyName` stores the *parameter* name and `ModelMetadata.ContainerType` is `null`.

The quickest fix for your 2.1 Preview1 app is to change
``` c#
if (!context.Metadata.IsComplexType && !string.IsNullOrEmpty(context.Metadata.PropertyName))
```
to
``` c#
if (!context.Metadata.IsComplexType && context.Metadata.MetadataKind == ModelMetadataKind.Property)
```

**Comment by dougbu** [comment]
And. @stevejgordon you're correct that our `ModelMetadata.DebuggerToString()` makes a similar assumption about non-`ModelMetadataKind.Type` metadata. It will hit an NRE for any `ModelMetadataKind.Parameter` metadata.

**Comment by stevejgordon** [comment]
Thanks @dougbu and @mkArtakMSFT for investigating so quickly.

Is this likely to change before RTM? I can imagine it's possible that quite a few people are using similar code which may break in these circumstances. If this is to be the recommended workaround I'll include details in my upcoming blog post :-)

**Comment by dougbu** [comment]
@stevejgordon the target milestone for a fix is `2.1.0-preview2`. That's not a firm commitment but it is a description of our intent.

A fix will entail some or all of the following
- Add a `ParameterName` property to use instead of `PropertyName` when it's `ModelMetadataKind.Parameter` metadata or add a `Name` property that is non-`null` for anything but `ModelMetadataKind.Type` metadata
  - That is, restore the previous "meaning" of `PropertyName` going forward
- Improve documentation of `ModelMetadata` to clarify when `ContainerMetadata`, `ContainerType`, `PropertyName`, `ParameterName` or `Name`, and related properties are or are not `null`
- Add a compatibility switch to maintain the `ModelMetadata` and top-level validation behaviour when running in "2.0 mode"
- Correct places in ASP.NET Core MVC where we make incorrect assumptions about `ModelMetadata` properties
  - `ModelMetadata.DebuggerToString()`, as you noted
  - `ModelBinderFactory.Key.ToString()`, also used only when debugging an MVC application
  - `ComplexTypeModelBinder.CreateModel(…)` where the logging of an obscure case *may* throw because `ContainerType == null`

**Comment by dougbu** [comment]
@mkArtakMSFT cost is on the S / M border

**Comment by dougbu** [comment]
Part 1:
- 5e245da326 in `release/2.1`
- 79c4986685 merge to `dev`

**Comment by dougbu** [comment]
fc3a815e57 in `release/2.1`
24eaa740f5 merge to `dev`
